File: src/wheel/llm/imp/deepseek_engine.py
import json
import urllib.error
import urllib.request

from src.wheel.llm.llm_engine import LLMEngine
from src.wheel.llm.llm_models import LLMConfig
from src.wheel.logger import get_logger
from src.wheel.scheduler.scheduler_factory import SchedulerFactory


class DeepSeekEngine(LLMEngine):
    _BASE_URL = "https://api.deepseek.com/v1"

    def __init__(self, config: LLMConfig):
        self._config = config
        self._scheduler = SchedulerFactory().create(None)
        self._log = get_logger("llm")
        self._base_url = config.base_url or self._BASE_URL

    async def start(self) -> None:
        await self._scheduler.start()
        self._log.info("LLM started: model={}", self._config.model)

    async def stop(self) -> None:
        await self._scheduler.stop()

    async def chat(
        self, messages: list[dict], tools: list[dict] | None = None
    ) -> dict:
        return await self._scheduler.run_blocking(
            self._chat_sync, messages, tools
        )

    def _chat_sync(
        self, messages: list[dict], tools: list[dict] | None
    ) -> dict:
        body: dict = {
            "model": self._config.model,
            "messages": messages,
            "temperature": self._config.temperature,
            "max_tokens": self._config.max_tokens,
        }
        if tools:
            body["tools"] = tools

        body_json = json.dumps(body, ensure_ascii=False)
        self._log.info("chat request → model={} msgs={} tools={} body={}",
                        self._config.model, len(messages),
                        len(tools) if tools else 0, body_json[:500])

        req = urllib.request.Request(
            f"{self._base_url}/chat/completions",
            data=body_json.encode("utf-8"),
            headers={
                "Content-Type": "application/json",
                "Authorization": f"Bearer {self._config.api_key}",
            },
        )

        try:
            with urllib.request.urlopen(req, timeout=self._config.timeout) as resp:
                raw = resp.read().decode("utf-8")
                data = json.loads(raw)
                self._log.info("chat response ← status={} body={}",
                               resp.status, raw[:500])
        except urllib.error.HTTPError as e:
            body_text = ""
            try:
                body_text = e.read().decode("utf-8", errors="replace")
            except Exception:
                pass
            self._log.error("chat error ← status={} body={}", e.code, body_text[:300])
            raise RuntimeError(
                f"LLM API returned {e.code}: {body_text[:200]}"
            ) from e
        except urllib.error.URLError as e:
            self._log.error("chat error ← reason={}", e.reason)
            raise RuntimeError(f"LLM API unreachable: {e.reason}") from e

        choice = data["choices"][0]
        msg = choice["message"]
        result: dict = {"role": "assistant", "content": msg.get("content") or ""}
        if msg.get("tool_calls"):
            result["tool_calls"] = msg["tool_calls"]
        return result

    def health_check(self) -> bool:
        try:
            req = urllib.request.Request(
                f"{self._base_url}/models",
                headers={
                    "Authorization": f"Bearer {self._config.api_key}"
                },
            )
            with urllib.request.urlopen(req, timeout=10) as resp:
                return 200 <= resp.status < 500
        except Exception:
            return False

    @property
    def name(self) -> str:
        return f"llm({self._config.provider}:{self._config.model})"

File: lynne-cpp/src/wheel/llm/imp/deepseek_engine.cpp
#include "wheel/llm/imp/deepseek_engine.h"
#include "wheel/scheduler/scheduler_factory.h"

#include <httplib.h>

namespace lynne {
namespace wheel {

static const char* kDefaultBaseUrl = "https://api.deepseek.com/v1";

DeepSeekEngine::DeepSeekEngine(const LLMConfig& config)
    : config_(config) {
    scheduler_ = SchedulerFactory().create();
}

DeepSeekEngine::~DeepSeekEngine() {
    stop();
    delete scheduler_;
}

std::string DeepSeekEngine::name() const {
    std::string n = "llm(" + config_.provider + ":" + config_.model + ")";
    return n;
}

void DeepSeekEngine::start() {
    if (started_) return;
    scheduler_->start();
    started_ = true;
}

void DeepSeekEngine::stop() {
    if (!started_) return;
    scheduler_->stop();
    scheduler_->step();
    started_ = false;
}

bool DeepSeekEngine::health_check() {
    return started_;
}

void DeepSeekEngine::chat(
    const nlohmann::json& messages,
    std::function<void(nlohmann::json)> on_result,
    std::function<void(const std::string&)> on_error,
    const nlohmann::json& tools
) {
    auto base_url = config_.base_url.empty() ? kDefaultBaseUrl : config_.base_url;

    std::string host;
    int port;
    std::string path_prefix;
    if (!parse_base_url(base_url, host, port, path_prefix)) {
        if (on_error) on_error("invalid base_url: " + base_url);
        return;
    }

    nlohmann::json body;
    body["model"] = config_.model;
    body["messages"] = messages;
    body["temperature"] = config_.temperature;
    body["max_tokens"] = config_.max_tokens;
    if (!tools.is_null()) {
        body["tools"] = tools;
    }
    std::string body_str = body.dump();

    auto ctx = std::make_shared<ChatCtx>();
    ctx->host = host;
    ctx->port = port;
    ctx->path = path_prefix + "/chat/completions";
    ctx->body = body_str;
    ctx->api_key = config_.api_key;
    ctx->timeout_sec = config_.timeout;

    scheduler_->run_blocking(
        [ctx]() {
            try {
                httplib::SSLClient cli(ctx->host, ctx->port);
                cli.set_connection_timeout(ctx->timeout_sec, 0);
                cli.set_read_timeout(ctx->timeout_sec, 0);
                cli.set_write_timeout(ctx->timeout_sec, 0);

                httplib::Headers headers = {
                    {"Content-Type", "application/json"},
                    {"Authorization", "Bearer " + ctx->api_key},
                };

                auto res = cli.Post(ctx->path, headers, ctx->body, "application/json");

                if (res) {
                    if (res->status == 200) {
                        ctx->result = nlohmann::json::parse(res->body);
                        ctx->ok = true;
                    } else {
                        std::string err_body = res->body.substr(0, 300);
                        ctx->error = "LLM API returned " + std::to_string(res->status)
                                   + ": " + err_body;
                        ctx->ok = false;
                    }
                } else {
                    auto err = res.error();
                    ctx->error = "LLM API connection failed: "
                               + httplib::to_string(err);
                    ctx->ok = false;
                }
            } catch (const std::exception& e) {
                ctx->error = "LLM exception: " + std::string(e.what());
                ctx->ok = false;
            }
        },
        [ctx, on_result, on_error]() {
            if (ctx->ok && on_result) {
                on_result(ctx->result);
            } else if (!ctx->ok && on_error) {
                on_error(ctx->error);
            }
        }
    );
}

bool DeepSeekEngine::parse_base_url(const std::string& url,
                                     std::string& host, int& port, std::string& path) {
    std::string s = url;
    size_t scheme_len = 0;
    if (s.rfind("https://", 0) == 0) {
        scheme_len = 8;
        port = 443;
    } else if (s.rfind("http://", 0) == 0) {
        scheme_len = 7;
        port = 80;
    } else {
        return false;
    }
    s = s.substr(scheme_len);

    auto slash = s.find('/');
    std::string host_port;
    if (slash == std::string::npos) {
        host_port = s;
        path = "";
    } else {
        host_port = s.substr(0, slash);
        path = s.substr(slash);
    }

    auto colon = host_port.find(':');
    if (colon == std::string::npos) {
        host = host_port;
    } else {
        host = host_port.substr(0, colon);
        port = std::stoi(host_port.substr(colon + 1));
    }
    return true;
}

void DeepSeekEngine::step() {
    scheduler_->step();
}

void DeepSeekEngine::run() {
    scheduler_->run();
}

} // namespace wheel
} // namespace lynne
